org.eclipse.sprotty/src/main/java/org/eclipse/sprotty/DefaultDiagramServer.java

```diff
@@ -263,31 +269,14 @@ public void setNeedsClientLayout(boolean value) {
 		this.needsClientLayout = value;
 	}
 	
-	/**
-	 * Whether and when the server needs to compute the layout of parts of the model. The layout is computed
-	 * with the layout engine configured with {@link #setLayoutEngine(ILayoutEngine)}, so returning {@code true}
-	 * here makes sense only if such an engine is available.
-	 *
-	 * <p>The default implementation returns the value configured with {@link #setServerLayoutKind(ServerLayoutKind)},
-	 * but this can be overridden to determine the value depending on the given model. The initial value
-	 * is {@code false}.</p>
-	 */
-	protected ServerLayoutKind getServerLayoutKind(SModelRoot root) {
-		return this.serverLayoutKind;
-	}
-	
-	public void setServerLayoutKind(ServerLayoutKind value) {
-		this.serverLayoutKind = value;
-	}
-	
 	/**
 	 * Submit a new or updated model to the client. If client layout is required, a {@link RequestBoundsAction}
 	 * is sent, otherwise either a {@link SetModelAction} or an {@link UpdateModelAction} is sent depending on
 	 * the {@code update} parameter.
 	 */
 	protected CompletableFuture<Void> submitModel(SModelRoot newRoot, boolean update, Action cause) {
 		if (needsClientLayout(newRoot)) {
-			if (getServerLayoutKind(newRoot) == ServerLayoutKind.NONE) {
+			if (!getLayoutEngine().needsServerLayout(newRoot, cause)) {
```

Comment:
We used to check for `null` services defensively (see `listener` variable three lines below). Shall we change that? Then we should do it consistently.

Reply:
But I'm more inclined to keep the defensive null checks.